### PR TITLE
Cooperative settle

### PR DESCRIPTION
**Pull request by loredanacirstea**
fixes https://github.com/raiden-network/raiden-contracts/issues/49

Adds support for an instant channel close & settle if both participants exchange signatures on balances calculated off-chain.

### CODE REVIEW

**Comment by LefterisJP** [comment]
Starting review of this.

File: raiden_contracts/contracts/TokenNetwork.sol [hunk 1]
```diff
@@ -766,6 +827,31 @@ contract TokenNetwork is Utils {
         signature_address = ECVerify.ecverify(message_hash, non_closing_signature);
     }
 
+    function recoverAddressFromCooperativeSettleSignature(
```
Comment (LefterisJP) [inline raiden_contracts/contracts/TokenNetwork.sol]: Spec question: Shouldn't the format of this signature be in the spec?

File: raiden_contracts/tests/test_channel_cooperative_settle.py [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+
+    balance_A_fail1 = 20
+    balance_B_fail1 = 11
+    balance_A_fail2 = 6
+    balance_B_fail2 = 8
+
+    # Create channel and deposit
+    channel_identifier = create_channel_and_deposit(A, B, deposit_A, deposit_B)
+
+    # Create cooperative settle signatures
```
Comment (LefterisJP) [inline raiden_contracts/tests/test_channel_cooperative_settle.py]: pedantic: No reason for the comment as the function is very aptly named :)

File: raiden_contracts/contracts/TokenNetwork.sol [hunk 3]
```diff
@@ -766,6 +827,31 @@ contract TokenNetwork is Utils {
...
+        uint256 channel_identifier,
+        address participant1,
+        uint256 participant1_balance,
+        address participant2,
```
Comment (LefterisJP) [inline raiden_contracts/contracts/TokenNetwork.sol]: From chat: 

> some context about me adding the addresses in the signature (not only the balance values) is - only having the balance values there in an unenforced order, can make it confusing to determine/check whose is the balance.

Why would it make it confusing? First balance would be participant1's balance and second balance participant2 balance.

File: raiden_contracts/contracts/TokenNetwork.sol (hunk 1 above)
Comment (loredanacirstea) [reply raiden_contracts/contracts/TokenNetwork.sol]: Yes. I will make a spec PR after this is merged. This was something that we did not discuss before.

File: raiden_contracts/tests/test_channel_cooperative_settle.py (hunk 2 above)
Comment (loredanacirstea) [reply raiden_contracts/tests/test_channel_cooperative_settle.py]: Yes, removing them. The comments were there first. Then came the rest.

File: raiden_contracts/contracts/TokenNetwork.sol (hunk 3 above)
Comment (loredanacirstea) [reply raiden_contracts/contracts/TokenNetwork.sol]: Yes, but in the Raiden client side: - you need to choose an order first and then send it to the other participant to sign. This is were I see issues - you have to pay extra care that the node receiving this message to sign knows how to interpret the balances correctly. 
I think this format brings more clarity.

Comment (LefterisJP) [reply raiden_contracts/contracts/TokenNetwork.sol]: > This is were I see issues - you have to pay extra care that the node receiving this message to sign knows how to interpret the balances correctly.

Each node will know its balance and the balance of its counterparty so I still don't see the scenario where having the addresses in the signature is really required.

Would like to hear @hackaugusto's opinion on this too.

> I think this format brings more clarity.

True, but this signature is intended to be made and consumed by a raiden client, so human readable clarity is not a goal here.

Comment (LefterisJP) [reply raiden_contracts/contracts/TokenNetwork.sol]: I agree. Changes to the signature if needed can happen in another PR. Whenever @hackaugusto becomes available we can see what his opinion is.

Comment (hackaugusto) [reply raiden_contracts/contracts/TokenNetwork.sol]: I'm fine with either. having the participant addresses is a bit more explicit wich is nice

**Pull request merged**